File: src/CubeBraidSDK/CameraSDK/CameraSDK_demo.cpp
#include <iostream>
#include <iomanip>
#include "CubeBraidSDK/CameraSDK/CameraSDK.h" // 替换为你实际的头文件名称

int main()
{
    std::cout << "=== Camera3D C++ SDK Demo 启动 ===" << std::endl;
    camera3d_sdk::Camera3D camera;

    // 配置相机参数
    std::string camera_ip = "192.168.23.203"; 

    // 准备算法输入参数
    // 手眼标定参数示例
    camera3d_sdk::CalibrationPose calib_pose;
    calib_pose.x = 0.107419f;
    calib_pose.y = 1.03611f;
    calib_pose.z = 0.23078f;
    calib_pose.qw = -0.4595f;
    calib_pose.qx = 0.595359f;
    calib_pose.qy = -0.519468f;
    calib_pose.qz = 0.405645f;

    // 常用 AGV / 模式参数
    const float agv_x = 1.643f;                    // AGV 前方距离
    const float agv_y = 0.387f;                    // AGV 左侧距离
    const float angle = 0.3f;                     // 倾角仪角度
    const float j1_angle = 15.0f;                 // 一轴关节角
    const bool integrated_mode = false;          // false: 摆台模式, true: 装卸一体模式

    // ------------------------------------------------------------
    // 调用算法接口 1: 集装箱内部/斜坡基准点计算 (processTradition)
    // ------------------------------------------------------------
    std::cout << "\n--- 执行：集装箱内/斜坡基准点计算 ---" << std::endl;
    camera3d_sdk::Point3D tradition_result;
    camera.processTradition(
        calib_pose,
        camera_ip,
        0,                 // model_mod: 0-第一面顶吸基准点, 1-其它面
        agv_x,
        agv_y,
        angle,
        j1_angle,
        integrated_mode,
        tradition_result
    );

    std::cout << std::fixed << std::setprecision(3);
    std::cout << "[SUCCESS] 传统基准点计算结果 -> X: " << tradition_result.x 
                << " m, Y: " << tradition_result.y 
                << " m, Z: " << tradition_result.z << " m" << std::endl;

    // ------------------------------------------------------------
    // 调用算法接口 2: 最后一面侧吸基准点计算 (processLastSurface)
    // ------------------------------------------------------------
    std::cout << "\n--- 执行：最后一面侧吸基准点计算 ---" << std::endl;
    camera3d_sdk::Point3D last_surface_result;
    camera.processLastSurface(
        calib_pose,
        camera_ip,
        agv_x,
        agv_y,
        j1_angle,
        integrated_mode,
        last_surface_result
    );

    std::cout << "[SUCCESS] 最后一面基准点计算结果 -> X: " << last_surface_result.x 
        << " m, Y: " << last_surface_result.y 
        << " m, Z: " << last_surface_result.z << " m" << std::endl;

    // ------------------------------------------------------------
    // 调用算法接口 3: 航向角偏差计算 (processYaw)
    // ------------------------------------------------------------
    // 配置相机参数
    std::string camera_ip_up = "192.168.23.203"; 

    // 手眼标定参数示例
    camera3d_sdk::CalibrationPose calib_pose_up;
    calib_pose_up.x = 0.107419f;
    calib_pose_up.y = 1.03611f;
    calib_pose_up.z = 0.23078f;
    calib_pose_up.qw = -0.4595f;
    calib_pose_up.qx = 0.595359f;
    calib_pose_up.qy = -0.519468f;
    calib_pose_up.qz = 0.405645f;

    std::cout << "\n--- 执行：AGV 航向角偏差计算 ---" << std::endl;
    double yaw_bias = 0.0;
    camera.processYaw(
        calib_pose_up,
        camera_ip_up,
        agv_x,
        agv_y,
        j1_angle,
        integrated_mode,
        yaw_bias
    );

    std::cout << std::setprecision(4);
    std::cout << "[SUCCESS] 偏航角偏差 (Yaw): " << yaw_bias << " 度" << std::endl;

    return 0;
}

File: scripts/CameraSDK/camera_sdk_demo.py
from camera_sdk import Camera3DSDK, Camera3DCalibrationPose

def main():
    print("=== Camera3D Python SDK Demo 程序启动 ===")

    # 定义测试用的相机 IP
    target_camera_ip = "192.168.23.203"
    target_camera_ip_up = "192.168.23.88"

    # 配置手眼标定外参位姿 (相机在机器人机械臂/车体上的相对姿态)
    calib_pose = Camera3DCalibrationPose(
        x=0.12, y=-0.05, z=0.85,
        qw=1.0, qx=0.0, qy=0.0, qz=0.0
    )

    # 使用 contextmanager (with 语句) 管理 SDK 上下文生命周期，防止内存泄漏
    try:
        with Camera3DSDK() as camera:
            # --------------------------------------------------------
            # 算法测试 1: 集装箱内部/斜坡基准点计算
            # --------------------------------------------------------
            pt_tradition = camera.process_tradition(
                pose=calib_pose,
                camera_ip=target_camera_ip,  # 补全 camera_ip 参数
                model_mod=0,
                agv_x=1.25,
                agv_y=0.85,
                angle=0.3,
                j1_angle=15.0,
                integrated_mode=False
            )
            print(f"[成功] 传统基准点坐标: {pt_tradition}")

            # --------------------------------------------------------
            # 算法测试 2: 最后一面侧吸基准点计算
            # --------------------------------------------------------
            pt_last = camera.process_last_surface(
                pose=calib_pose,
                camera_ip=target_camera_ip,  # 补全 camera_ip 参数
                agv_x=1.25,
                agv_y=0.85,
                j1_angle=15.0,
                integrated_mode=False
            )
            print(f"[成功] 最后一面基准点坐标: {pt_last}")

            # --------------------------------------------------------
            # 算法测试 3: AGV 航向角偏差计算
            # --------------------------------------------------------
            yaw_deg = camera.process_yaw(
                pose=calib_pose,
                camera_ip=target_camera_ip_up,  # 补全 target_camera_ip_up 参数
                slam_x=1.25,
                slam_y=0.85,
                j1_angle=15.0,
                integrated_mode=False
            )
            print(f"[成功] 航向角偏差 (Yaw): {yaw_deg:.4f} 度")
       

    except RuntimeError as e:
        print(f"[运行时错误] {e}")
    except Exception as e:
        print(f"[未知异常] {e}")

    print("=== Demo 退出 ===")

if __name__ == "__main__":
    main()